projects/http-auth/main.go

```diff
@@ -140,5 +153,107 @@ func main() {
 		fmt.Fprintf(w, "Internal server error")
 	})
 
+	http.HandleFunc("/authenticated", func(w http.ResponseWriter, r *http.Request) {
+		// get the authorization header
+		authorizationHeader := r.Header.Get("Authorization")
+		// fmt.Println("authorizationHeader", authorizationHeader)
+
+		// there was no authentication header
+		if len(authorizationHeader) == 0 {
+			w.Header().Set("WWW-Authenticate", `Basic realm=""`)
+			w.WriteHeader(http.StatusUnauthorized)
+			fmt.Fprintf(w, "401 Unauthorized [no authorization header]")
+			return
+		}
+
+		// authorizationHeader header should be in the format "Basic <base64 encoded payload>"
+		authorizationString := strings.Split(authorizationHeader, " ")
+		// fmt.Println("authorizationString", authorizationString)
+
+		// must be 2 composite parts and the first must be "Basic"
+		if len(authorizationString) != 2 || authorizationString[0] != "Basic" {
+			w.WriteHeader(http.StatusBadRequest)
+			fmt.Fprintf(w, "400 Bad Request [malformed authorization header]")
+			return
+		}
+
+		// decode the base64 encoded payload
+		payload, err := base64.StdEncoding.DecodeString(authorizationString[1])
+		// fmt.Println("payload", payload, "err", err)
+
+		// the payload couldn't be decoded
+		if err != nil {
+			w.WriteHeader(http.StatusBadRequest)
+			fmt.Fprintf(w, "400 Bad Request [couldn't decode the authorization header payload]")
+			return
+		}
+
+		// split the payload into it's composite parts
+		credentials := strings.Split(string(payload), ":")
+		// fmt.Println("credentials", credentials)
+
+		username := credentials[0]
+		password := credentials[1]
+		// fmt.Println("username", username, "password", password)
+
+		// username/password must not be empty
+		if len(username) == 0 || len(password) == 0 {
+			w.WriteHeader(http.StatusBadRequest)
+			fmt.Fprintf(w, "400 Bad Request [either username or password were missing]")
+			return
+		}
+
+		// get the username/password from the .env file
+		usernameFromEnv := os.Getenv("AUTH_USERNAME")
+		passwordFromEnv := os.Getenv("AUTH_PASSWORD")
+
+		// see if the username and password match from the .env
+		if username != usernameFromEnv || password != passwordFromEnv {
+			w.WriteHeader(http.StatusUnauthorized)
+			fmt.Fprintf(w, "401 Unauthorized [either username or password were not correct]")
+			return
+		}
+
+		// if we reach here the authorization was successful
+		w.Header().Add("Content-Type", "text/html")
+		fmt.Fprintf(w, "<!DOCTYPE html>\n<html>\n<p>Hello %s!</p>\n", username)
```

Comment:
Might this need escaping, as the query parameters in the previous stage did?

Reply:
yes, i missed that